src/ray/common/client_connection.h

```diff
@@ -193,11 +214,105 @@ class ClientConnection : public ServerConnection<T> {
   std::vector<uint8_t> read_message_;
 };
 
+/// \typename SimpleConnection
+///
+/// A generic type representing a client connection on a server. In addition to
+/// writing messages to the client, like in ServerConnection, this typename can
+/// also be used to process messages asynchronously from client.
+template <typename T>
+class SimpleConnection {
+ public:
+  SimpleConnection(boost::asio::basic_stream_socket<T> &&socket)
+      : socket_(std::move(socket)) {}
+
+  /// \return The ClientID of the remote client.
+  const ClientID &GetClientId() const { return client_id_; }
+
+  /// \param client_id The ClientID of the remote client.
+  void SetClientID(const ClientID &client_id) { client_id_ = client_id; }
+
+  /// Read a message from the client. This method is NOT thread-safe.
+  ///
+  /// \param type The message type (e.g., a flatbuffer enum).
+  /// \param message A pointer to the message buffer.
+  /// \return Status.
+  ray::Status ReadMessage(MessageType type, std::unique_ptr<uint8_t[]> &message);
```

Comment:
It looks like the ReadMessage and WriteMessage without lock won't be used in any cases currently, how about remove them from public interface to avoid confusion?

Reply:
And if you rename this class to `ThreadSafeConnection`, then you can also drop the ThreadSafe suffix.